Apps/FlightSoftware/Watchdog/src/i2c_sensors.c

```diff
@@ -0,0 +1,564 @@
+// i2c_sensors.c
+
+#include "include/i2c_sensors.h"
+#include "include/i2c.h"
+
+#include <string.h>
+
+//###########################################################
+// Private definitions and globals
+//###########################################################
+
+// Some definitions to identify magic numbers in the code
+#define BOOL short
+static const BOOL TRUE = 1;
+static const BOOL FALSE = 0;
+
+static INS_Sensors__InternalState internals = { 0 };
+
+//###########################################################
+// Private function declarations
+//###########################################################
+
+BOOL I2C_Sensors__readRegNonBlocking(uint8_t devAddr,
+                                     uint8_t regAddr,
+                                     uint8_t nackMaskBit,
+                                     GaugeReadingState nextState,
+                                     void* output,
+                                     BOOL* done,
+                                     BOOL* gotOutput);
+
+
+void I2C_Sensors__writeRegNonBlocking(uint8_t devAddr,
+                                      uint8_t regAddr,
+                                      uint8_t data,
+                                      BOOL* done,
+                                      BOOL* success);
+
+BOOL I2C_Sensors__chargeLsb();
+BOOL I2C_Sensors__chargeMsb();
+BOOL I2C_Sensors__voltageLsb();
+BOOL I2C_Sensors__voltageMsb();
+BOOL I2C_Sensors__currentLsb();
+BOOL I2C_Sensors__currentMsb();
+BOOL I2C_Sensors__gaugeTempLsb();
+BOOL I2C_Sensors__gaugeTempMsb();
+
+//###########################################################
+// Public function definitions
+//###########################################################
+void I2C_Sensors__init()
+{
+    // Just init the generic I2C module we depend on
+    I2C__init();
+}
+
+void I2C_Sensors__initiateGaugeReadings()
+{
+    internals.gState = GRS__CHARGE_LSB;
+    internals.readings.nackMask = 0U;
+}
+
+I2C_Sensors__Status
+I2C_Sensors__getGaugeReadingStatus(I2C_Sensors__Readings* readings)
+{
+    if (NULL == readings) {
+        return I2C_SENSORS__STATUS__ERROR__NULL;
+    }
+
+    switch (internals.gState) {
+        case GRS__UNKNOWN:
+            return I2C_SENSORS__STATUS__ERROR__READINGS_NOT_STARTED;
+
+        case GRS__CHARGE_LSB:
+        case GRS__CHARGE_MSB:
+        case GRS__VOLTAGE_LSB:
+        case GRS__VOLTAGE_MSB:
+        case GRS__CURRENT_LSB:
+        case GRS__CURRENT_MSB:
+        case GRS__GAUGE_TEMP_LSB:
+        case GRS__GAUGE_TEMP_MSB:
+            return I2C_SENSORS__STATUS__INCOMPLETE;
+
+        case GRS__DONE:
+            memcpy(readings, &internals.readings, sizeof(*readings));
```

Comment:
Thanks! I stumbled upon it when I was trying to run the code and all the i2c data was offset (i.e. charge values in voltage array, voltage values in current array, etc), I imagine because of the presence of `gState` in the `INS_Sensors__InternalState` struct